File: aoc/2024/aoc202116.hs
module AOC202116 where

import Control.Monad.State(State,evalState,get,put)

import AOC

aoc = AOC {
    day="../../2021/input/16",
    aocTests=[
        AOCTest {
            testData="8A004A801A8002F478",
            testResult=Just "16",
            testResult2=Nothing
            },
        AOCTest {
            testData="620080001611562C8802118E34",
            testResult=Just "12",
            testResult2=Nothing
            },
        AOCTest {
            testData="C0015000016115A2E0802F182340",
            testResult=Just "23",
            testResult2=Nothing
            },
        AOCTest {
            testData="A0016C880162017C3686B18A3D4780",
            testResult=Just "31",
            testResult2=Nothing
            },
        AOCTest {
            testData="C200B40A82",
            testResult=Nothing,
            testResult2=Just "3"
            },
        AOCTest {
            testData="04005AC33890",
            testResult=Nothing,
            testResult2=Just "54"
            },
        AOCTest {
            testData="880086C3E88112",
            testResult=Nothing,
            testResult2=Just "7"
            },
        AOCTest {
            testData="CE00C43D881120",
            testResult=Nothing,
            testResult2=Just "9"
            },
        AOCTest {
            testData="D8005AC2A8F0",
            testResult=Nothing,
            testResult2=Just "1"
            },
        AOCTest {
            testData="F600BC2D8F",
            testResult=Nothing,
            testResult2=Just "0"
            },
        AOCTest {
            testData="9C005AC2F8F0",
            testResult=Nothing,
            testResult2=Just "0"
            },
        AOCTest {
            testData="9C0141080250320F1802104A08",
            testResult=Nothing,
            testResult2=Just "1"
            }
        ],
    aocCode=Code {
        codeParse=parse,
        codeParse2=parse,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

data Packet = Literal Int Int | Operator Int Int [Packet] deriving Show

parse = evalState parsePacket . parseHex

parseHex :: String -> [Int]
parseHex = concatMap p
  where
    p '0' = [0,0,0,0]
    p '1' = [0,0,0,1]
    p '2' = [0,0,1,0]
    p '3' = [0,0,1,1]
    p '4' = [0,1,0,0]
    p '5' = [0,1,0,1]
    p '6' = [0,1,1,0]
    p '7' = [0,1,1,1]
    p '8' = [1,0,0,0]
    p '9' = [1,0,0,1]
    p 'A' = [1,0,1,0]
    p 'B' = [1,0,1,1]
    p 'C' = [1,1,0,0]
    p 'D' = [1,1,0,1]
    p 'E' = [1,1,1,0]
    p 'F' = [1,1,1,1]

parsePacket :: State [Int] Packet
parsePacket = do
    version <- parseNumber 3
    typeID <- parseNumber 3
    if typeID == 4
      then do
        value <- parseLiteral 0
        return $ Literal version value
      else do
        subpackets <- parseSubpackets
        return $ Operator version typeID subpackets

parseNumber :: Int -> State [Int] Int
parseNumber nbits = do
    bits <- get
    let (numberBits,rest) = splitAt nbits bits
    put rest
    return $ evalBits 0 numberBits
  where
    evalBits accumulator [] = accumulator
    evalBits accumulator (bit:bits) = evalBits (accumulator*2+bit) bits

parseLiteral :: Int -> State [Int] Int
parseLiteral accumulator = do
    (flag:rest) <- get
    put rest
    n <- parseNumber 4
    if flag == 1
      then parseLiteral (16*accumulator+n)
      else return $ (16*accumulator+n)

parseSubpackets :: State [Int] [Packet]
parseSubpackets = do
    lengthTypeID <- parseNumber 1
    if lengthTypeID == 0
      then do
        nbits <- parseNumber 15
        bits <- get
        let (subpacketBits,rest) = splitAt nbits bits
        put subpacketBits
        subpackets <- p0 []
        put rest
        return subpackets
      else do
        npackets <- parseNumber 11
        p1 npackets []
  where
    p0 accumulator = do
        bits <- get
        if null bits
          then return $ reverse accumulator
          else do
            packet <- parsePacket
            p0 (packet:accumulator)

    p1 0 accumulator = return $ reverse accumulator
    p1 n accumulator = do
        packet <- parsePacket
        p1 (n-1) (packet:accumulator)

result (Literal version _) = version
result (Operator version _ subpackets) = version + sum (map result subpackets)

greaterthan :: [Int] -> Int
greaterthan [a,b] | a > b = 1 | otherwise = 0

lessthan :: [Int] -> Int
lessthan [a,b] | a < b = 1 | otherwise = 0

equalto :: [Int] -> Int
equalto [a,b] | a == b = 1 | otherwise = 0

result2 (Literal _ n) = n
result2 (Operator _ 0 subpackets) = sum $ map result2 subpackets
result2 (Operator _ 1 subpackets) = product $ map result2 subpackets
result2 (Operator _ 2 subpackets) = minimum $ map result2 subpackets
result2 (Operator _ 3 subpackets) = maximum $ map result2 subpackets
result2 (Operator _ 5 subpackets) = greaterthan $ map result2 subpackets
result2 (Operator _ 6 subpackets) = lessthan $ map result2 subpackets
result2 (Operator _ 7 subpackets) = equalto $ map result2 subpackets

File: aoc/2025/aoc202116.go
package main

import (
	"math"
)

func init() {
	Register(&aoc202116{
		AOC: AOC{
			Day:           16,
			InputFilename: "../2021/input/16.txt",
			Tests: []Test{
				Test{
					Input: `8A004A801A8002F478`,
					Part1: "16",
					Part2: "",
				},
				Test{
					Input: `620080001611562C8802118E34`,
					Part1: "12",
					Part2: "",
				},
				Test{
					Input: `C0015000016115A2E0802F182340`,
					Part1: "23",
					Part2: "",
				},
				Test{
					Input: `A0016C880162017C3686B18A3D4780`,
					Part1: "31",
					Part2: "",
				},
				Test{
					Input: `C200B40A82`,
					Part1: "",
					Part2: "3",
				},
				Test{
					Input: `04005AC33890`,
					Part1: "",
					Part2: "54",
				},
				Test{
					Input: `880086C3E88112`,
					Part1: "",
					Part2: "7",
				},
				Test{
					Input: `CE00C43D881120`,
					Part1: "",
					Part2: "9",
				},
				Test{
					Input: `D8005AC2A8F0`,
					Part1: "",
					Part2: "1",
				},
				Test{
					Input: `F600BC2D8F`,
					Part1: "",
					Part2: "0",
				},
				Test{
					Input: `9C005AC2F8F0`,
					Part1: "",
					Part2: "0",
				},
				Test{
					Input: `9C0141080250320F1802104A08`,
					Part1: "",
					Part2: "1",
				},
			},
		},
	})
}

type aoc202116 struct {
	AOC
}

func (aoc *aoc202116) bits(input *Input) []bool {
	bits := []bool{}
	for ch := range input.Chars() {
		if ch >= 'A' {
			ch -= 'A' - 10
		} else if ch >= '0' {
			ch -= '0'
		} else {
			return bits
		}
		for bit := byte(8); bit > 0; bit >>= 1 {
			bits = append(bits, bit&ch != 0)
		}
	}
	return bits
}

func (aoc *aoc202116) readInt(bits []bool, length int) (int, []bool, bool) {
	i := 0
	for range length {
		if len(bits) == 0 {
			return 0, nil, false
		}
		i <<= 1
		if bits[0] {
			i |= 1
		}
		bits = bits[1:]
	}
	return i, bits, true
}

func (aoc *aoc202116) parse1(bits []bool) (int, []bool, bool) {
	versions, bits, ok := aoc.readInt(bits, 3)
	if !ok {
		return 0, nil, false
	}
	typeID, bits, _ := aoc.readInt(bits, 3)
	switch typeID {
	case 4:
		for {
			val, bits2, ok := aoc.readInt(bits, 5)
			if !ok {
				return 0, nil, false
			}
			bits = bits2
			if val&16 == 0 {
				break
			}
		}
	default:
		lengthTypeID, bits2, ok := aoc.readInt(bits, 1)
		if !ok {
			return 0, nil, false
		}
		bits = bits2
		switch lengthTypeID {
		case 0:
			length, bits2, _ := aoc.readInt(bits, 15)
			bits = bits2[length:]
			bits2 = bits2[:length]
			for {
				v, bits3, ok := aoc.parse1(bits2)
				bits2 = bits3
				if !ok {
					break
				}
				versions += v
			}
		case 1:
			count, bits2, _ := aoc.readInt(bits, 11)
			bits = bits2
			for range count {
				v, bits2, _ := aoc.parse1(bits)
				bits = bits2
				versions += v
			}
		}
	}
	return versions, bits, true
}

func (aoc *aoc202116) Part1(input *Input) string {
	result, _, ok := aoc.parse1(aoc.bits(input))
	if !ok {
		panic("bad input")
	}
	return IntResult(result)
}

func (aoc *aoc202116) parse2(bits []bool) (int, []bool, bool) {
	_, bits, ok := aoc.readInt(bits, 3)
	if !ok {
		return 0, nil, false
	}
	typeID, bits, _ := aoc.readInt(bits, 3)
	switch typeID {
	case 4:
		v := 0
		for {
			val, bits2, ok := aoc.readInt(bits, 5)
			if !ok {
				return 0, nil, false
			}
			v <<= 4
			v |= val & 15
			bits = bits2
			if val&16 == 0 {
				return v, bits, true
			}
		}
	default:
		lengthTypeID, bits2, ok := aoc.readInt(bits, 1)
		if !ok {
			return 0, nil, false
		}
		bits = bits2
		operands := []int{}
		switch lengthTypeID {
		case 0:
			length, bits2, _ := aoc.readInt(bits, 15)
			bits = bits2[length:]
			bits2 = bits2[:length]
			for {
				val, bits3, ok := aoc.parse2(bits2)
				bits2 = bits3
				if !ok {
					break
				}
				operands = append(operands, val)
			}
		case 1:
			count, bits2, _ := aoc.readInt(bits, 11)
			bits = bits2
			for range count {
				val, bits2, _ := aoc.parse2(bits)
				bits = bits2
				operands = append(operands, val)
			}
		}
		var result int
		switch typeID {
		case 0:
			result = 0
			for _, val := range operands {
				result += val
			}
		case 1:
			result = 1
			for _, val := range operands {
				result *= val
			}
		case 2:
			result = math.MaxInt
			for _, val := range operands {
				result = min(result, val)
			}
		case 3:
			result = math.MinInt
			for _, val := range operands {
				result = max(result, val)
			}
		case 5:
			if operands[0] > operands[1] {
				result = 1
			}
		case 6:
			if operands[0] < operands[1] {
				result = 1
			}
		case 7:
			if operands[0] == operands[1] {
				result = 1
			}
		}
		return result, bits, true
	}
}

func (aoc *aoc202116) Part2(input *Input) string {
	result, _, ok := aoc.parse2(aoc.bits(input))
	if !ok {
		panic("bad input")
	}
	return IntResult(result)
}
